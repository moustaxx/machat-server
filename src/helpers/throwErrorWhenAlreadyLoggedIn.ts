import { ApolloError } from 'apollo-server-errors';

function throwErrorWhenAlreadyLoggedIn(isLoggedIn: boolean | null) {
    if (isLoggedIn) {
        throw new ApolloError('You are already logged in!', 'ALREADY_LOGGED_IN');
    }
}

export default throwErrorWhenAlreadyLoggedIn;
