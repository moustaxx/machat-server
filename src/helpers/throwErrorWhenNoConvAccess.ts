import { ForbiddenError } from 'apollo-server-errors';
import { TPrisma } from '../prismaClient';

const throwErrorWhenNoConvAccess = async (
    prisma: TPrisma,
    userID: number,
    conversationID: number,
): Promise<void> => {
    const data = await prisma.person.findUnique({
        select: {
            isAdmin: true,
            conversations: {
                select: {
                    id: true,
                },
            },
        },
        where: { id: userID },
    });

    const isParticipated = !!data?.conversations.find((conv) => {
        return conv.id === conversationID;
    });

    if (!data?.isAdmin && !isParticipated) {
        throw new ForbiddenError('Insufficient permissions');
    }
};

export default throwErrorWhenNoConvAccess;
