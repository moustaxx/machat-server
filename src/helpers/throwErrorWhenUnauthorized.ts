import { ApolloError } from 'apollo-server-errors';

function throwErrorWhenUnauthorized(clientID: number | null): asserts clientID is number {
    if (!clientID) {
        throw new ApolloError('You must be logged in!', 'UNAUTHORIZED');
    }
}

export default throwErrorWhenUnauthorized;
