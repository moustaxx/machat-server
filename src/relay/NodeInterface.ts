import { Field, ID, InterfaceType } from 'type-graphql';
import { ConversationType, MessageType, PersonType } from '../modules';

@InterfaceType({
    isAbstract: true,
    description: 'An object with a global ID.',
    resolveType: (value) => {
        if ('createdAt' in value && 'name' in value) return ConversationType;
        if ('authorID' in value) return MessageType;
        if ('username' in value) return PersonType;
        throw new Error('Unknown type');
    },
})
export abstract class Node {
    @Field(() => ID, {
        name: 'id',
        description: 'The global ID of the object.',
    })
    readonly globalId!: string;
}
