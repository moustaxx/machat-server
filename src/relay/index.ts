export * from './ConnectionArgs';
export * from './ConnectionType';
export * from './NodeInterface';
export * from './NodeResolver';
export * from './PageInfo';
export * from './EdgeType';
