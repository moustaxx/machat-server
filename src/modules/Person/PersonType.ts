/* eslint-disable @typescript-eslint/indent */
import { Conversation, Prisma } from '@prisma/client';
import { Args, Authorized, Ctx, Field, FieldResolver, ObjectType, Resolver, Root } from 'type-graphql';
import { ConnectionArguments, findManyCursorConnection } from '@devoxa/prisma-relay-cursor-connection';
import { ForbiddenError } from 'apollo-server-errors';

import { Context } from '../../context';
import { Message, Person, PersonMessagesArgs } from '../../generated/type-graphql';
import { ConversationConnection, ConversationType } from '../Conversation/ConversationType';
import { Connection, ConnectionArgs, ConnectionType, EdgeType, Node } from '../../relay';
import cursorUtils from '../../helpers/cursor';

@ObjectType({ implements: Node })
export class PersonType extends Node {
    id!: number;

    @Field((_type) => Boolean)
    isActive!: boolean;

    @Field((_type) => Date)
    createdAt!: Date;

    @Field((_type) => String)
    email!: string;

    @Field((_type) => Boolean)
    isAdmin!: boolean;

    @Field((_type) => Date, { nullable: true })
    lastSeen?: Date | null;

    @Field((_type) => String)
    username!: string;

    @Field((_type) => [Message])
    messages!: Message[];

    @Field((_type) => ConversationConnection)
    conversations!: Connection<ConversationType>;
}

@Resolver((_of) => PersonType)
export class PersonTypeResolver {
    @Authorized()
    @FieldResolver()
    isActive(@Root() person: PersonType, @Ctx() ctx: Context<true>) {
        return ctx.personActiveStatus.get(person.id);
    }

    @Authorized()
    @FieldResolver()
    async conversations(
        @Root() person: PersonType,
        @Args((_type) => ConnectionArgs) args: ConnectionArguments,
        @Ctx() { session, prisma }: Context<true>,
    ) {
        const myID = session.owner.id;
        const whereMe: Prisma.ConversationWhereInput = {
            participants: {
                some: {
                    id: myID,
                },
            },
        };

        const where: Prisma.ConversationWhereInput = (person.id === myID) ? whereMe : {
            AND: [
                whereMe,
                {
                    participants: {
                        some: {
                            id: person.id,
                        },
                    },
                },
            ],
        };


        return findManyCursorConnection<
            Conversation, Prisma.ConversationWhereUniqueInput
        >(
            async (convArgs) => prisma.conversation.findMany({ ...convArgs, where }),
            async () => prisma.conversation.count({ where }),
            args,
            cursorUtils,
        );
    }

    @FieldResolver((_type) => [Message])
    async messages(
        @Root() person: Person,
        @Ctx() { prisma, session }: Context,
        @Args() args: PersonMessagesArgs,
    ): Promise<Message[]> {
        if (!session?.owner?.isAdmin) {
            throw new ForbiddenError('Insufficient permissions');
        }

        return prisma.person.findUnique({
            where: { id: person.id },
            select: { messages: true },
        }).messages(args);
    }
}

@ObjectType()
export class PersonEdge extends EdgeType(PersonType) { }

@ObjectType()
export class PersonConnection extends ConnectionType(PersonEdge) { }
