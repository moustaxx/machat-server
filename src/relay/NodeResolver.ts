/* eslint-disable @typescript-eslint/indent */
import { Arg, Ctx, FieldResolver, ID, Info, Query, Resolver, Root } from 'type-graphql';

import { Node } from './NodeInterface';
import { Connection } from './ConnectionType';
import { Context } from '../context';
import { PersonResolver, ConversationResolver } from '../modules';
import throwErrorWhenUnauthorized from '../helpers/throwErrorWhenUnauthorized';

export type TNodeModel<TModel> = Omit<TModel, 'id' | 'globalId'> & {
    id: `${string}:${string}`;
};

export type TNodeConnection<TModel> = Connection<TNodeModel<TModel>>;

export const toGlobalId = (typename: string, id: string | number) => {
    return `${typename}:${id}`;
};

export const fromGlobalId = (nodeID: string) => {
    const delimiterPos = nodeID.indexOf(':');
    const id = Number(nodeID.substring(delimiterPos + 1));
    const typename = nodeID.substring(0, delimiterPos);

    return { id, typename };
};

@Resolver(() => Node)
export class NodeResolver {
    @FieldResolver()
    globalId(
        @Root() { id }: { id: string | number },
        @Info() { parentType: { name } }: { parentType: { name: string } },
    ): string {
        return toGlobalId(name, id);
    }

    private async fetcher(globalId: string, ctx: Context | Context<true>) {
        throwErrorWhenUnauthorized(ctx.session);
        const { typename, id } = fromGlobalId(globalId);

        switch (typename) {
            case 'ConversationType': {
                return ConversationResolver.prototype.conversation(
                    { whereId: id },
                    ctx as Context<true>,
                );
            }
            case 'PersonType': {
                return PersonResolver.prototype.person({ id }, ctx as Context<true>);
            }
            case 'MessageType': {
                throw new Error('Unallowed type');
            }
            default: {
                throw new Error('Unknown type');
            }
        }
    }

    @Query(() => Node, {
        nullable: true,
        description: 'Fetches an object given its global ID.',
    })
    async node(
        @Arg('id', () => ID, { description: 'The global ID of the object.' })
        globalId: string,
        @Ctx() ctx: Context,
    ): ReturnType<NodeResolver['fetcher']> {
        return this.fetcher(globalId, ctx);
    }

    @Query(() => [Node], {
        nullable: 'items',
        description: 'Fetches objects given their global IDs.',
    })
    nodes(
        @Arg('ids', () => [ID], { description: 'The global IDs of the objects.' })
        globalIds: Array<string>,
        @Ctx() ctx: Context,
    ): Array<ReturnType<NodeResolver['fetcher']>> {
        return globalIds.map(async (id) => this.fetcher(id, ctx));
    }
}
