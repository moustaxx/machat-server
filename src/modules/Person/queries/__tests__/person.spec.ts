import { initTestServer, ITestUtils } from '../../../../tests/helpers';
import { TNodeModel } from '../../../../relay';
import { PersonType } from '../../PersonType';

let t: ITestUtils;

beforeAll(async () => {
    t = await initTestServer();
});

afterAll(async () => {
    await t.closeTestServer();
});

const queryStringFull = `
    query person($whereId: Int!) {
        person(where: { id: $whereId }) {
            id
            username
            email
            isActive
            isAdmin
            lastSeen
            createdAt
            conversations(first: 20) {
                edges {
                    node {
                        id
                    }
                }
            }
        }
    }
`;

const queryString = `
    query person($whereId: Int!) {
        person(where: { id: $whereId }) {
            id
        }
    }
`;

it('should return person', async () => {
    const { cookies } = await t.createRandomUserAndLogin({ isAdmin: true });
    const { user } = await t.createRandomUser();

    type TPerson = { person: TNodeModel<PersonType> };
    const { data } = await t.gqlQuery<TPerson>({
        query: queryStringFull,
        variables: { whereId: user.id },
        cookies,
    });

    expect(data).toMatchSnapshot({
        person: {
            email: expect.any(String),
            username: expect.any(String),
            createdAt: expect.any(String),
        },
    });

    expect(data.person).toMatchObject({
        email: user.email,
        username: user.username,
        createdAt: user.createdAt.toISOString(),
    });
});


it('should throw error when user not found', async () => {
    const { cookies } = await t.createRandomUserAndLogin();

    const { errors } = await t.gqlQuery({
        query: queryString,
        variables: { whereId: 999 },
        cookies,
    });

    const errorCode = errors?.[0].extensions?.code;
    expect(errorCode).toEqual('USER_NOT_FOUND');
});
