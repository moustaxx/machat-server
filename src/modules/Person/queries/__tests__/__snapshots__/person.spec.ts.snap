// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return person 1`] = `
Object {
  "person": Object {
    "conversations": Object {
      "edges": Array [],
    },
    "createdAt": Any<String>,
    "email": Any<String>,
    "id": "PersonType:2",
    "isActive": false,
    "isAdmin": false,
    "lastSeen": null,
    "username": Any<String>,
  },
}
`;
